Sancus step spy module

Sven-C opened the pull request:
Sancus-step now includes a debug module for automatic configuration of nemesis parameters, such as isr_reti_latency (latency between the end of the reti instruction and the first new instruction executed by the module) and sm_exit_latency (the latency between the arrival of an interrupt and the execution of the isr)

Sven-C committed:
Clear TAR upon completing a trace

Sven-C committed:
Fixed warning for printf being undefined in sancus_step.c. Also added a function that just returns the isr latency instead of printing it.

Sven-C committed:
Redefined some C macro's, so that inline assembly code does not have to be touched if anything needs to be changed

Sven-C committed:
Fix delay change -- needs better fix though

Sven-C committed:
WIP

Sven-C committed:
Prototype of sancus-step debug/spy module

Sven-C committed:
Merge branch 'master' into sancus-step-debug-module

Sven-C committed:
Added a macro for easier definition of an isr for timer related interrupts

Sven-C committed:
Sancus-step debug module v1.0

Sven-C committed:
Added isr stackpointer to timer module and improved macro to switch stacks if necessary

Sven-C committed:
Sancus step debug module  version 1.1
Includes automatic discovery of exit latency

jovanbulck commented on include/sancus_support/sancus_step.h:
```diff
@@ -3,50 +3,101 @@
+void sancus_step_init(void);
+void sancus_step_end(void);
+
+// sancus step configuration parameters
```
to avoid unintended name collisions in linked code, I propose you stick consistently to the `__ss` name space for all global variable names in the sancus-step library as you already did in some places?

jovanbulck commented on include/sancus_support/sancus_step.h:
```diff
@@ -3,50 +3,101 @@
-            "2: ; cont after if-then-else               \n\t"   \
-            "reti                                       \n\t"   \
-            ::"m"(TAR),"m"(TACCR0),"m"(TACTL):);
+    if (isr_interrupted_sm)
```
is this commented code still needed? If not, let's remove it for clarity

jovanbulck commented on src/sancus-step/sancus_step.c:
```diff
@@ -1,11 +1,96 @@
+__attribute__((naked))
+int SM_ENTRY(ssdbg) ss_dbg_get_info(void)
+{
+    /*
```
let's also remove this commented code for clarity and just add comments between the asm where the IRQ is supposed to arrive?

jovanbulck commented on src/sancus-step/sancus_step.c:
```diff
@@ -1,11 +1,96 @@
+    return __ss_isr_tar_entry - sm_exit_latency;
+}
+
+int ss_dbg_entry_delay = 0;
```
cf, `__ss` globale variable name namespace comment

jovanbulck commented on include/sancus_support/timer.h:
```diff
@@ -25,4 +31,21 @@ int  timer_tsc_end(void);
+{                                                                   \
+    __asm__ __volatile__(                                           \
+            "cmp #0x0, r1\n\t"                                      \
+            "jne 1f\n\t"                                            \
```
don't you have to save caller-save register R15 etc here before jumping to the ISR (since the ISR function is allowed to overwrite them and the hardware only pushed R1/R2 for an unprotected interrupt)?

jovanbulck commented on include/sancus_support/timer.h:
```diff
@@ -1,13 +1,19 @@
-#define TACTL_DISABLE       (TACLR)
-#define TACTL_ENABLE        (TASSEL_2 + MC_1 + TAIE)
-#define TACTL_CONTINUOUS    ((TASSEL_2 + MC_2) & ~TAIE)
+#define TACTL_DISABLE       (TACLR) // 0x04
```
it should be possible to include these symbolic constants in your inline asm as "constant" contraint parameter ("i" or something I believe). That would be preferable to avoid magic numbers in the inline asm :)

jovanbulck commented on src/sancus-step/sancus_step.c:
```diff
@@ -1,11 +1,96 @@
+    __asm__ __volatile__(
+                "mov &%0, &ss_dbg_entry_delay\n\t"
+                "mov &%0, r15\n\t"
+                "mov #0x4, &%1\n\t"
```
cf., asm as "constant" contraint parameter comment above

jovanbulck commented on include/sancus_support/sancus_step.h:
```diff
@@ -3,50 +3,101 @@
+void timerA_isr_entry(void)                                         \
+{                                                                   \
+    __asm__("mov &%0, &%2; save tar\n\t"                            \
+            "mov #0x4, &%4; disable timer\n\t "                     \
```
cf., asm as "constant" contraint parameter comment above

jovanbulck reviewed (changes requested):
Thanks for the code! :+1: I added a few (minor) comments mostly regarding code style to further improve this before merging :)

Sven-C committed:
Fixed macro for Timer ISR

Sven-C committed:
Sancus-step debug module v1.1.1

All sancus-step related variables live in __ss_ namespace
Used 'i' option in extended asm to remove magic constants from asm code
Moved comments about asm code into the code itself

Sven-C committed:
Restored missing constant definition

Sven-C committed:
Some extra comments on where interrupts should arrive in the sancus-step debug module

jovanbulck commented:
looks good, thanks!
